src/components/vm/VmInitializer.tsx

```diff
@@ -124,7 +126,31 @@ export default function VmInitializer() {
             const cleanProps = mapValues({ to, href, ...rest }, (val: any, key: string) => {
               if (!['to', 'href'].includes(key)) return val;
               if (key === 'href' && !val) val = to;
-              return typeof val === 'string' && isValidAttribute('a', 'href', val) ? val : 'about:blank';
+              const isAtrValid = typeof val === 'string' && isValidAttribute('a', 'href', val);
+              let linkValue;
+              if (isAtrValid) {
+                if (val.startsWith('?')) {
+                  const currentParams = new URLSearchParams(
+                    Object.entries(router.query).reduce((acc, [key, value]) => {
+                      if (typeof value === 'string') {
+                        acc[key] = value;
+                      }
+                      return acc;
+                    }, {} as { [key: string]: string }), // Add index signature to the object type
+                  );
+                  const newParams = new URLSearchParams(val);
+                  newParams.forEach((value, key) => {
+                    currentParams.set(key, value);
+                  });
+
+                  linkValue = `${router.pathname}?${currentParams.toString()}`;
```

Comment:
@shelegdmitriy Are we sure `router.pathname` is what we want here? I feel like that would still have the same issue that we were seeing before. We might want to consider using `router.asPath`. Here is an example from a different Next JS App when logging `router`:

```ts
{
  asPath: "/events/2/edit",
  pathname: "/events/[eventId]/edit"
}
```

Notice how `pathname` still has the URL segment names with square brackets instead of the actual URL value. Maybe I'm missing something here, but just wanted to double check. :)

Reply:
Well, let me explain a bit (maybe I missed something):

At first, this change affects links which starts with `?` sign, for example:
`<Link href="?path=some-path">some link</Link>`

Let's say we have some widget with url
`http://localhost:3000/some-account.near/widget/TestComponent`
and there is a link to some content (page or section) with such url:
`?path=some-path` (usually we're trying to put a full url instead of structure like this, as an example ProfilePage). In this case `"val"="?path=some-path"`

Normally, by clicking on link `?path=some-path` I'm expecting to see page with this url:
`http://localhost:3000/some-account.near/widget/TestComponent?path=some-path`
This is a valid url and correct behavior. But if we take a look at `router.asPath` and `router.pathname` we'll see this:

```jsx
{
  asPath: "some-account.near/widget/TestComponent?path=some-path",
  pathname: "[componentAccountId]/widget/[componentName]"
}
```

Back to the question Caleb, if I put a `asPath` instead of `pathname` here:

`linkValue = ${router.asPath}?${currentParams.toString()};`

or throw away code from line **133** and to **146** and put this line of code:

`linkValue = ${router.asPath}${val};`

this will brake the href by adding `val` at the end of the string every time I click on any link. We don't need to simply add a `val` (meaning `?path=some-path`) instead, we have to split previous path with the new one. And that's exactly what I've been trying to implement.

Hope my explanation helps 🙂 Am I missed something?

Reply:
Also you can try to click on some links using a preview link because this PR doesn't require to launch a component's repo separately

Reply:
@calebjacob let me know what do you think 🙂

Reply:
@shelegdmitriy Thanks for the detailed response! My main source of confusion is coming from Next JS - its behavior was really confusing me since I could have sworn it handled links like this natively:  `<Link href="?foo=bar" />`. I did a bunch of digging to help understand what's going on.

If your route is a static path without any dynamic segments (eg: `/some-folder/some-file`), it can handle `href="?foo=bar"` just fine.

However, if your path has any dynamic segments (eg: `/products/123` - the `123` mapping to something like `[productId]`) - you have to include all of those dynamic segments as part of your query like so: `href="?foo=bar&productId=123"`. If you do that, it will work as expected. However, passing only `href="?foo=bar` makes Next JS throw an error because it requires `productId` to be passed. It's a bit confusing due to how Next JS treats dynamic URL segments and query params as the same concept.

So all this means is that technically we only need to do this:

```linkValue = `?${currentParams.toString()}`;```

Prepending the pathname isn't really necessary, but it doesn't hurt anything:

```linkValue = `${router.pathname}?${currentParams.toString()}`;```

Your logic works well because you end up with this string output: `/[componentAccountId]/widget/[componentName]?foo=bar&componentAccountId=alem-lib.near&componentName=Index`

But this string output would also work fine: `?foo=bar&componentAccountId=alem-lib.near&componentName=Index`

I might have been the only one confused by this, haha. It was helpful to understand Next JS's behavior for myself. This looks good! Let's merge it 👍

Reply:
@calebjacob thanks for the explanation! Yes, you're right? your code works the same so I'm good with your solution 🙂 Thank you so much!